xds/internal/balancer/edsbalancer/xds_client_wrapper.go

```diff
@@ -184,44 +148,26 @@ func (c *xdsClientWrapper) replaceXDSClient(newClient xdsClientInterface, newBal
 // the balancerName (from bootstrap file or from service config) changed.
 // - if balancer names are the same, do nothing, and return false
 // - if balancer names are different, create new one, and return true
-func (c *xdsClientWrapper) updateXDSClient(config *EDSConfig, attr *attributes.Attributes) (bool, error) {
-	if attr != nil {
-		if clientFromAttr, _ := attr.Value(xdsinternal.XDSClientID).(xdsClientInterface); clientFromAttr != nil {
-			// This will also clear balancerName, to indicate that client is
-			// from attributes.
-			return c.replaceXDSClient(clientFromAttr, ""), nil
-		}
+func (c *xdsClientWrapper) updateXDSClient(attr *attributes.Attributes) (bool, error) {
+	if attr == nil {
+		return false, fmt.Errorf("unexported nil attributes, want attributes with xdsClient")
 	}
-
-	clientConfig, err := bootstrapConfigNew()
-	if err != nil {
-		// TODO: propagate this error to ClientConn, and fail RPCs if necessary.
-		clientConfig = &bootstrap.Config{BalancerName: config.BalancerName}
+	// TODO: change the way xdsClient is retrieved from attributes. One option
+	// is to add helper functions.
+	//
+	// Or, since xdsClient will become a singleton, this can just call
+	// xdsclient.New() instead. And if we decide to do this, do it in Build
+	// instead of when handling updates.
+	clientFromAttr, _ := attr.Value(xdsinternal.XDSClientID).(xdsClientInterface)
+	if clientFromAttr == nil {
+		return false, fmt.Errorf("no xdsClient found in attributes")
 	}
 
-	if c.balancerName == clientConfig.BalancerName {
+	if c.xdsClient == clientFromAttr {
 		return false, nil
 	}
-
-	var dopts []grpc.DialOption
-	if dialer := c.bbo.Dialer; dialer != nil {
-		dopts = []grpc.DialOption{grpc.WithContextDialer(dialer)}
-	}
-
-	// TODO: there's no longer a need to read bootstrap file and create a new
-	// xds client. The EDS balancer should always get the xds client from
-	// attributes. Otherwise, this function should just fail. Also, xdsclient
-	// will be shared by multiple clients, so trying to make an xds client is
-	// just the wrong move.
-	newClient, err := xdsclientNew(xdsclient.Options{Config: *clientConfig, DialOpts: dopts})
-	if err != nil {
-		// This should never fail. xdsclientnew does a non-blocking dial, and
-		// all the config passed in should be validated.
-		//
-		// This could leave c.xdsClient as nil if this is the first update.
-		return false, fmt.Errorf("eds: failed to create xdsClient, error: %v", err)
-	}
-	return c.replaceXDSClient(newClient, clientConfig.BalancerName), nil
+	c.xdsClient = clientFromAttr
```

Comment:
Under what circumstances will the `client`  sent via attributes change? Given that a resolver makes an `xdsClient` at build time, do we even need to support the case of this attribute changing?

Reply:
Right, this should never change.

I'm now more inclined to just call `xdsclient.New()` in each balancer's Build() now (or, if build is too early, when the balancer needs to make the first xds request). That would simplify these things.